### PR TITLE
UpToDateCheck string allocation reduction

### PR DESCRIPTION
**Pull request by drewnoakes**
Relates to #3744.

Probably easiest to review commit-by-commit:

- a340a4d9c4cfbd8876770e8d74b3b5d24d544527 defers the splitting of the `MSBuildAllProjects` property value until it is used, reducing retained memory and preventing the majority of these split values being promoted from Gen0
- 882d5a067161a814bd258a1b70cdfdd41a929e53 adds another item to `Delimiter` to reduce params array allocations
- 7b729cc156bf63179254c85e4bf87574d6448b07 introduces `LazyStringSplit` and unit tests, supporting splitting strings lazily, and with considerably fewer allocated bytes than `string.Split`
- 496708c7a5fefad1fb925c79255966b5b6033a74 uses `LazyStringSplit` to split the `MSBuildAllProjects` value into short-lived strings, and also prevents allocating redundant strings that exist after one that identifies the build is out of date

I ran some benchmarks to verify behaviour of `LazyStringSplit` using the value of `MSBuildAllProjects` observed when building `Roslyn.sln`:

```
BenchmarkDotNet=v0.11.1, OS=Windows 10.0.17134.228 (1803/April2018Update/Redstone4)
Intel Core i7-6700HQ CPU 2.60GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
  [Host]       : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3132.0
  Clr          : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.3132.0
  LegacyJitX86 : .NET Framework 4.7.2 (CLR 4.0.30319.42000), 32bit LegacyJIT-v4.7.3132.0
Runtime=Clr
```

|    Method |          Job |       Jit | Platform |      Mean |     Error |    StdDev | Rank |   Gen 0 | Allocated |
|---------- |------------- |---------- |--------- |----------:|----------:|----------:|-----:|--------:|----------:|
|     Split |          Clr |    RyuJit |      X64 | 13.489 us | 0.2454 us | 0.2410 us |    3 | 11.1694 |   35265 B |
| LazySplit |          Clr |    RyuJit |      X64 |  7.359 us | 0.1086 us | 0.1016 us |    1 |  3.9825 |   12553 B |
|     Split | LegacyJitX86 | LegacyJit |      X86 | 16.707 us | 0.3169 us | 0.3523 us |    4 | 10.8643 |   34242 B |
| LazySplit | LegacyJitX86 | LegacyJit |      X86 | 10.134 us | 0.1881 us | 0.1760 us |    2 |  3.7384 |   11789 B |

`Roslyn.sln` has 164 projects, so the reduction in the _Allocated_ column can be multiplied out to around 3.7MB.

The _Gen0_ column is the number of collections per 1,000 invocations of the operation.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs [hunk 1]
```diff
@@ -119,11 +119,7 @@ private void OnProjectChanged(IProjectSubscriptionUpdate e)
             _msBuildProjectFullPath = e.CurrentState.GetPropertyOrDefault(ConfigurationGeneral.SchemaName, ConfigurationGeneral.MSBuildProjectFullPathProperty, _msBuildProjectFullPath);
             _msBuildProjectDirectory = e.CurrentState.GetPropertyOrDefault(ConfigurationGeneral.SchemaName, ConfigurationGeneral.MSBuildProjectDirectoryProperty, _msBuildProjectDirectory);
             _outputRelativeOrFullPath = e.CurrentState.GetPropertyOrDefault(ConfigurationGeneral.SchemaName, ConfigurationGeneral.OutputPathProperty, _outputRelativeOrFullPath);
-
-            string[] allProjects = e.CurrentState.GetPropertyOrDefault(ConfigurationGeneral.SchemaName, ConfigurationGeneral.MSBuildAllProjectsProperty, string.Empty)
-                .Split(new[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
-            _imports.Clear();
-            _imports.AddRange(allProjects);
+            _msBuildAllProjects = e.CurrentState.GetPropertyOrDefault(ConfigurationGeneral.SchemaName, ConfigurationGeneral.MSBuildAllProjectsProperty, "");
```
Comment (Pilchie) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: @davkean does this repo have a stated preference for `string.Empty`/`String.Empty`/`""`?

**Comment by Pilchie** [comment]
Nice!

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Text/LazyStringSplitTests.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        [InlineData("", ';', new string[0])]
+        [InlineData(";", ';', new string[0])]
+        [InlineData(";;", ';', new string[0])]
+        [InlineData(";;;", ';', new string[0])]
+        [InlineData(";;;a", ';', new[] { "a" })]
+        [InlineData("a;;;", ';', new[] { "a" })]
+        [InlineData(";a;;", ';', new[] { "a" })]
+        [InlineData(";;a;", ';', new[] { "a" })]
+        [InlineData("a", ';', new[] { "a" })]
+        [InlineData("aa", ';', new[] { "aa" })]
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Text/LazyStringSplitTests.cs]: Nit, I find it a little easier to view inputs/expectations if you turn this into table

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    ///     methods on <see cref="string"/>.
+    /// </remarks>
+    internal readonly struct LazyStringSplit : IEnumerable<string>
+    {
+        private readonly string _input;
+        private readonly char _delimiter;
+
+        public LazyStringSplit(string input, char delimiter)
+        {
+            _input = input ?? throw new ArgumentNullException(nameof(input));
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs]: We use Requires for contract checking.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                        _index++;
+                        continue;
+                    }
+
+                    Current = _input.Substring(_index, length);
+                    _index = delimiterIndex + 1;
+                    return true;
+                }
+
+                return false;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs]: Just as a small note of interest, String.Split returns a 1 element array on empty string, you return zero elements, this is behavor acts more like the SplitReturningEmptyIfEmpty extension method we already have.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Text/LazyStringSplitTests.cs (hunk 2 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Text/LazyStringSplitTests.cs]: Fixed in 189d881616fe3dfead1d3861b6ab2645559251d8.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs (hunk 3 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs]: Fixed in e678f4d4d2edfbb17cf687b64cb24d7caba0173d.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs (hunk 4 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/Text/LazyStringSplit.cs]: This type mimics the behaviour of `string.Split` when passing `StringSplitOptions.RemoveEmptyEntries`, which returns an empty array for on empty string.

I added an equivalence check to the unit test, and updated the documentation in 889f0d9c4cdd998d88f5ee60311321ce84a642e5.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: Find in files shows:

- 94 instances of `string.Empty`
- 119 instances of `String.Empty`
- 2575 instances of `""` (probably lots of false positives)

I don't feel strongly either way, though tend to use `""`.

Comment (davidwengier) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: String 🤮

Its the wrong colour! How can anyone live like this :P

Comment (jmarolf) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: I vote `string.Empty` please

Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: Sure. I won't change any more of these.

Comment (Pilchie) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: There are people who think that you should use `string` as a type name for variables, but `String` any time you access a static member off of the type.  I'm not sure *why* people think that, but some do.  🤷‍♂️

**Pull request merged**